pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -21,50 +36,491 @@ func init() {
+	policy.mutex.Lock()
+	defer policy.mutex.Unlock()
+
+	input := map[string]interface{}{
+		"name":        "mount_overlay",
+		"containerID": containerID,
+		"layerPaths":  layerPaths,
+	}
+	result, err := policy.Query(input)
+	if err != nil {
+		return err
+	}
+
+	if !result.Allowed() {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("overlay mount not allowed by policy.\ninput: %s", string(input_json))
+	}
+
+	// we store the mapping of container ID -> layerPaths for later
+	// use in EnforceCreateContainerPolicy here.
+	containerMap := policy.data["containers"].(map[string]interface{})
+	if _, found := containerMap[containerID]; found {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("container %s already mounted.\ninput: %s", containerID, string(input_json))
+	}
+
+	containerMap[containerID] = map[string]interface{}{
+		"containerID": containerID,
+		"layerPaths":  layerPaths,
+	}
+	return nil
+}
+
+// Rego does not have a way to determine the OS path separator
+// so we append it via these methods.
+func sandboxMountsDir(sandboxID string) string {
+	return fmt.Sprintf("%s%c", spec.SandboxMountsDir(sandboxID), os.PathSeparator)
+}
+
+func hugePagesMountsDir(sandboxID string) string {
+	return fmt.Sprintf("%s%c", spec.HugePagesMountsDir(sandboxID), os.PathSeparator)
+}
+
+func (policy *RegoEnforcer) EnforceCreateContainerPolicy(
+	sandboxID string,
+	containerID string,
+	argList []string,
+	envList []string,
+	workingDir string,
+	mounts []oci.Mount,
+) error {
+	policy.mutex.Lock()
+	defer policy.mutex.Unlock()
+
+	// first, we need to obtain the overlay filestytem information
+	// which was stored in EnforceOverlayMountPolicy
+	var containerInfo map[string]interface{}
+	containerMap := policy.data["containers"].(map[string]interface{})
+	if container, found := containerMap[containerID]; found {
+		containerInfo = container.(map[string]interface{})
+	} else {
+		return fmt.Errorf("container %s does not have a filesystem", containerID)
+	}
+
+	input := map[string]interface{}{
+		"name":         "create_container",
+		"argList":      argList,
+		"envList":      envList,
+		"workingDir":   workingDir,
+		"sandboxDir":   sandboxMountsDir(sandboxID),
+		"hugePagesDir": hugePagesMountsDir(sandboxID),
+		"mounts":       mounts,
+	}
+
+	// this adds the overlay layerPaths array to the input
+	for key, value := range containerInfo {
+		input[key] = value
+	}
+
+	result, err := policy.Query(input)
+	if err != nil {
+		return err
+	}
+
+	if result.Allowed() {
+		started := policy.data["started"].([]string)
+		policy.data["started"] = append(started, containerID)
+		containerInfo["argList"] = argList
+		containerInfo["envList"] = envList
+		containerInfo["workingDir"] = workingDir
+		return nil
+	} else {
```

Comment:
No need for the else as we early return no matter what on 470

Reply:
I personally find an explicit else to be clearer when scanning.

Reply:
Either works, my brain works in mysterious ways